### PR TITLE
Create TypeNameProcessingRegister

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
**What this PR does / why we need it**:

I have a need in `StorageTypeFactory` to track all of the `TypeName` references encountered, checking to see if we process a `TypeDefinition` for each one. I've factored this out as a utility type and intend to use it from at least two different locations. 

We'd brainstormed the name `DynamicTypeNameCheckList` for this type, but that didn't work quite so well in use. 

I've also changed the API for `TypeNameSet` to remove the returns from mutators, and cleaned up code that was reliant upon them.

**How does this PR make you feel**:
gif

**If applicable**:
- [x] this PR contains tests

### CODE REVIEW

File: hack/generator/pkg/astmodel/type_name_set.go [hunk 1]
```diff
@@ -14,23 +16,17 @@ type TypeNameSet map[TypeName]struct{}
...
-// Add includes the passed name in the set and returns the updated
-// set, so that adding can work for a nil set - this makes it more
-// convenient to add to sets kept in a map (in the way you might with
-// a map of slices).
-func (ts TypeNameSet) Add(val TypeName) TypeNameSet {
-	if ts == nil {
-		ts = make(TypeNameSet)
-	}
+// Add includes the passed name in the set
+func (ts TypeNameSet) Add(val TypeName) {
```
Comment (matthchr) [inline hack/generator/pkg/astmodel/type_name_set.go:28]: minor: The changes to `TypeNameSet` possibly deserve their own PR?

File: hack/generator/pkg/astmodel/type_name_processing_register_test.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+func Test_TypeNameProcessingRegister_Remove_NameIsNoLongerCompleted(t *testing.T) {
+	g := NewGomegaWithT(t)
+	register := NewTypeNameProcessingRegister()
+	register.Complete(gammaTypeName)
+	g.Expect(register.IsCompleted(gammaTypeName)).To(BeTrue())
+	g.Expect(register.IsPending(gammaTypeName)).To(BeFalse())
+	register.Remove(gammaTypeName)
+	g.Expect(register.IsCompleted(gammaTypeName)).To(BeFalse())
+	g.Expect(register.IsPending(gammaTypeName)).To(BeFalse())
+}
```
Comment (matthchr) [inline hack/generator/pkg/astmodel/type_name_processing_register_test.go:181]: minor: missing newline

File: hack/generator/pkg/astmodel/type_name_set_test.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	"testing"
+
+	. "github.com/onsi/gomega"
+)
+
+var (
+	oneTypeName = MakeTypeName(GenRuntimeReference, "One")
+	twoTypeName  = MakeTypeName(GenRuntimeReference, "Two")
+)
+
```
Comment (matthchr) [inline hack/generator/pkg/astmodel/type_name_set_test.go:13]: minor: Extra newline

File: hack/generator/pkg/astmodel/type_name_processing_register.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+func (reg *TypeNameProcessingRegister) Complete(name TypeName) {
+	if _, found := reg.known[name]; !found {
+		reg.known[name] = []string{}
+	}
+
+	reg.completed.Add(name)
+}
+
+// Reset marks the specified name as unprocessed (ie not complete)
+// As a side effect, the name will be added to the checklist if it is missing
```
Comment (matthchr) [inline hack/generator/pkg/astmodel/type_name_processing_register.go:44]: minor: What is "the checklist" - possibly prefer talking about it in terms of completed/known? I think you mean added to the list of completed types.

Comment (matthchr) [reply hack/generator/pkg/astmodel/type_name_processing_register.go:44]: oh, no "the checklist" is the types we know about but haven't necessarily confirmed yet?

File: hack/generator/pkg/astmodel/type_name_processing_register.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
+package astmodel
+
+// TypeNameProcessingRegister is a register of TypeNames that require processing, keeping track of which names have been
+// processed and which are still outstanding. This is used, for example, for a consistency check that all TypeName
+// references can be satisfied
+type TypeNameProcessingRegister struct {
```
Comment (matthchr) [inline hack/generator/pkg/astmodel/type_name_processing_register.go:6]: Do you have an example of how this is used? There are two things that I'm struggling with:

1. Does this belong in `astmodel` or is it actually specific to the storage processing? Trying to understand how it might be used outside of the storage type generation context, but first it would probably help to understand how it's being used in that context.
2. As discussed in person, it's not clear to me if `TypeWalker` can help you do this or not. Obviously the two things are very different - seeing how this was used (maybe in a draft PR?) might help me answer this question.

File: hack/generator/pkg/astmodel/type_name_set.go [hunk 6]
```diff
@@ -14,23 +16,17 @@ type TypeNameSet map[TypeName]struct{}
 // NewTypeNameSet makes a TypeNameSet containing the specified
 // names. If no elements are passed it might be nil.
```
Comment (babbageclunk) [inline hack/generator/pkg/astmodel/type_name_set.go:17]: Remove the sentence about it possibly being nil.

**Review by babbageclunk — APPROVED** [review]

**Comment by theunrepentantgeek** [comment]
No longer need this utility; other content was pulled out into PR #1607
